contracts/its-stacks-translator/src/contract/its_hub_message_to_clarity_bytes.rs

```diff
@@ -0,0 +1,342 @@
+
+use crate::error::ContractError;
+
+pub const MESSAGE_TYPE_INTERCHAIN_TRANSFER: u128 = 0;
+pub const MESSAGE_TYPE_DEPLOY_INTERCHAIN_TOKEN: u128 = 1;
+
+pub const CLARITY_NAME_TYPE: &str = "type";
+const CLARITY_NAME_SOURCE_CHAIN: &str = "source-chain";
+pub const CLARITY_NAME_TOKEN_ID: &str = "token-id";
+pub const CLARITY_NAME_SOURCE_ADDRESS: &str = "source-address";
+pub const CLARITY_NAME_DESTINATION_ADDRESS: &str = "destination-address";
+pub const CLARITY_NAME_AMOUNT: &str = "amount";
+pub const CLARITY_NAME_DATA: &str = "data";
+pub const CLARITY_NAME_NAME: &str = "name";
+pub const CLARITY_NAME_SYMBOL: &str = "symbol";
+pub const CLARITY_NAME_DECIMALS: &str = "decimals";
+const CLARITY_NAME_MINTER_BYTES: &str = "minter-bytes";
+
+pub fn its_hub_message_to_clarity_bytes(
+    its_hub_message: HubMessage,
+) -> Result<Vec<u8>, ContractError> {
+    let payload = match its_hub_message {
+        HubMessage::SendToHub { .. } => Err(ContractError::InvalidPayload),
+        HubMessage::ReceiveFromHub {
+            source_chain,
+            message,
+        } => match message {
+            Message::InterchainTransfer(InterchainTransfer {
+                token_id,
+                source_address,
+                destination_address,
+                amount,
+                data,
+            }) => get_its_interchain_transfer_payload_receive_from_hub(
+                source_chain,
+                token_id,
+                source_address.into(),
+                destination_address.into(),
+                amount,
+                data,
+            ),
+            Message::DeployInterchainToken(DeployInterchainToken {
+                token_id,
+                name,
+                symbol,
+                decimals,
+                minter,
+            }) => get_its_deploy_interchain_token_payload_receive_from_hub(
+                source_chain,
+                token_id,
+                name.into(),
+                symbol.into(),
+                decimals,
+                minter,
+            ),
+            Message::LinkToken(_) => Err(ContractError::InvalidPayload),
+        },
+        HubMessage::RegisterTokenMetadata(_) => Err(ContractError::InvalidPayload),
+    }?;
+
+    Ok(payload)
+}
+
+fn get_its_interchain_transfer_payload_receive_from_hub(
+    source_chain: ChainNameRaw,
+    token_id: TokenId,
+    source_address: HexBinary,
+    destination_address: HexBinary,
+    amount: Uint256,
+    data: Option<nonempty::HexBinary>,
+) -> Result<Vec<u8>, ContractError> {
+    let token_id: [u8; 32] = token_id.into();
+
+    if amount > Uint256::try_from(Uint128::MAX).map_err(|_| ContractError::InvalidAmount)? {
+        return Err(ContractError::InvalidAmount);
+    }
+
+    let amount: Uint128 = cosmwasm_std::Uint256::from(amount)
+        .try_into()
+        .map_err(|_| ContractError::InvalidAmount)?;
+
+    let tuple_data = TupleData::from_data(vec![
+        (
+            ClarityName::from(CLARITY_NAME_TYPE),
```

Comment:
do the field names need to be provided? Can't they be inferred from the order of encoding?

Reply:
The field names need to be provided, the actual order does not matter. That is the way the Clarity decoding works.